### PR TITLE
Avoid overflow

### PR DESCRIPTION
**Pull request by pacoyi**
System.out.println(new AtomicInteger(Integer.MAX_VALUE).incrementAndGet());
Output:
-2147483648

### CODE REVIEW

**Comment by stoneliyi** [comment]
this is a serious bug that bit us on production where more than 2 billion requests were made. and as it is the default rule provided by Spring Cloud, many applications could run into the issue. please merge it to fix the bug.

File: ribbon-loadbalancer/src/main/java/com/netflix/loadbalancer/AbstractServerPredicate.java [hunk 1]
```diff
@@ -137,6 +137,22 @@ public void setLoadBalancerStats(LoadBalancerStats stats) {
...
     }
+	
+	/**
+	 * Referenced from RoundRobinRule
+     * Inspired by the implementation of {@link AtomicInteger#incrementAndGet()}.
+     *
+     * @param modulo The modulo to bound the value of the counter.
+     * @return The next value.
+     */
+    private int incrementAndGetModulo(int modulo) {
```
Comment (qiangdavidliu) [inline ribbon-loadbalancer/src/main/java/com/netflix/loadbalancer/AbstractServerPredicate.java]: Hi, the potential busy spin here is a little concerning. What about something like
```
int current = Math.abs(nextIndex.incrementAndGet());
return current % modulo;
```

Comment (stoneliyi) [reply ribbon-loadbalancer/src/main/java/com/netflix/loadbalancer/AbstractServerPredicate.java]: as its comment suggests, it's the same as RoundRobinRule.incrementAndGetModulo(int modulo)(in master branch). it should be safe. 

and for "int current = Math.abs(nextIndex.incrementAndGet()); "

java docs writes "Note that if the argument is equal to the value of Integer.MIN_VALUE, the most negative representable int value, the result is that same value, which is negative."
https://docs.oracle.com/javase/7/docs/api/java/lang/Math.html#abs(int)

it seems that could be an issue when it's overflowed. 

@pacoyi is the one who made the PR though. :-)

Comment (qiangdavidliu) [reply ribbon-loadbalancer/src/main/java/com/netflix/loadbalancer/AbstractServerPredicate.java]: Fair point, that would indeed be an issue.

**Comment by stoneliyi** [comment]
thanks @pacoyi and @qiangdavidliu. 

in which ribbon release should we expect to see this fix?

**Comment by qiangdavidliu** [comment]
release 2.2.4 is out with this fix.

**Pull request merged**